kbfsgit/runner.go

```diff
@@ -232,56 +267,184 @@ func (r *runner) waitForJournal(ctx context.Context) error {
+		return err
+	}
+
+	for {
+		ref, err := refs.Next()
+		if errors.Cause(err) == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		value := ""
+		switch ref.Type() {
+		case plumbing.HashReference:
+			value = ref.Hash().String()
+		case plumbing.SymbolicReference:
+			value = "@" + ref.Target().String()
+		default:
+			value = "?"
+		}
+		refStr := value + " " + ref.Name().String() + "\n"
+		_, err = r.output.Write([]byte(refStr))
+		if err != nil {
+			return err
+		}
+	}
+
 	err = r.waitForJournal(ctx)
 	if err != nil {
 		return err
 	}
 	r.log.CDebugf(ctx, "Done waiting for journal")
 
-	// TODO(KBFS-2353/KBFS-2354): list actual references from the repo.
 	_, err = r.output.Write([]byte("\n"))
 	return err
 }
 
-func (r *runner) processCommands(ctx context.Context) error {
+// handleFetchBatch: From https://git-scm.com/docs/git-remote-helpers
+//
+// fetch <sha1> <name>
+// Fetches the given object, writing the necessary objects to the
+// database. Fetch commands are sent in a batch, one per line,
+// terminated with a blank line. Outputs a single blank line when all
+// fetch commands in the same batch are complete. Only objects which
+// were reported in the output of list with a sha1 may be fetched this
+// way.
+//
+// Optionally may output a lock <file> line indicating a file under
+// GIT_DIR/objects/pack which is keeping a pack until refs can be
+// suitably updated.
+func (r *runner) handleFetchBatch(ctx context.Context, args [][]string) (
+	err error) {
+	repo, err := r.initRepoIfNeeded(ctx)
+	if err != nil {
+		return err
+	}
+
+	r.log.CDebugf(ctx, "Fetching %d refs into %s", len(args), r.gitDir)
+
+	remote, err := repo.CreateRemote(&gogitcfg.RemoteConfig{
+		Name: localRepoRemoteName,
+		URL:  r.gitDir,
+	})
+
+	for _, fetch := range args {
+		if len(fetch) != 2 {
+			return errors.Errorf("Bad fetch request: %v", fetch)
+		}
+		refInBareRepo := fetch[1]
+
+		// Push into a local ref with a temporary name, because the
+		// git process that invoked us will get confused if we make a
+		// ref with the same name.  Later, delete this temporary ref.
+		localTempRef := plumbing.ReferenceName(refInBareRepo).Short() +
+			"-" + r.uniqID
+		refSpec := fmt.Sprintf(
+			"%s:refs/remotes/%s/%s", refInBareRepo, r.remote, localTempRef)
+		r.log.CDebugf(ctx, "Fetching %s", refSpec)
+
+		// Now "push" into the local repo to get it to store objects
+		// from the KBFS bare repo.
+		err = remote.Push(&gogit.PushOptions{
+			RemoteName: localRepoRemoteName,
+			RefSpecs:   []gogitcfg.RefSpec{gogitcfg.RefSpec(refSpec)},
+		})
+		if err != nil && err != gogit.NoErrAlreadyUpToDate {
+			return err
+		}
+
+		// Delete the temporary refspec now that the objects are
+		// safely stored in the local repo.
+		refSpec = fmt.Sprintf(":refs/remotes/%s/%s", r.remote, localTempRef)
+		err = remote.Push(&gogit.PushOptions{
+			RemoteName: localRepoRemoteName,
+			RefSpecs:   []gogitcfg.RefSpec{gogitcfg.RefSpec(refSpec)},
+		})
```

Comment:
Assuming `fetch[2]` is the local ref name for the fetch (or "remote" ref name in reverse speaking), I was under the impression that we'd need to  create a ref under that name in the on-disk repo. Then I saw the comment above:

> fetching directly via the runner doesn't leave any refs, those would normally be created by the `git` process that invokes the runner

Is this what the comment is talking about? Does it mean `fetch[2]` is useless then?

Reply:
As above, `fetch[1]` is the name of the reference the user wants to fetch from the bare repo.  `fetch[0]` is the sha1 hash, which I'm not using here.

The comment is trying to say that calling `Push` on the remote (the local disk in this case) doesn't actually create any ref in the remote, it only populates the objects.  The `git` command line process later creates the ref to be whatever the user specified it should be (which is not passed to the GRH).  In the test, where that comment was taken from, we're not using the command line, so no ref is created, which means we have to check out using the sha1 directly.